python/psim/plugins.py

```diff
@@ -0,0 +1,55 @@
+"""Defines the plugin interface a basic set of plugins for the end user.
+
+Plugins are a simple way of a user tacking on extra functionality to a
+simulation. Simply inherit from this class, override the default callback
+functions, and register them to a simulation runner.
+"""
+
+class Plugin(object):
+    """Defines the basic interface all simulation plugins should adhere to.
+    """
+    def __init__():
+        super(Plugin, self).__init__()
+
+    def initialize(self, sim):
+        """Pluging function called by the simulation before taking the first
+        step.
+
+        This could be used to change initial conditions or something similiar.
+        """
+        pass
+
+    def prestep(self, sim):
+        """Plugin function called by the simulation prior to every step.
+        """
+        pass
+
+    def poststep(self, sim):
+        """Plugin function called by the simulation right after every step.
+        """
+        pass
+
+    def cleanup(self, sim):
+        """Plugin function called by the simulation just before exiting.
+        """
+        pass
+
+
+class RunForNSteps(Plugin):
+    """Simulation plugin that terminates a simulation after it has taken a
+    a specified number of steps.
+    """
+    def __init__(self, n):
+        super(Plugin, self).__init__()
+
+        self._n = n
+        self._steps = 0
+
+    def poststep(self, sim):
+        super(RunForNSteps, self).poststep(sim)
+
+        if self._steps >= self.n
+            sim.should_stop()
+            return
+
+        self._steps = self._steps + 1
```

Comment:
Here a _very_ simple example of what a plugin could do. Something that would be a little more complicated is a plugin acting as the attitude estimator and one acting as the attitude controller.